src/Streamly/Internal/Data/Parse/Types.hs

```diff
@@ -0,0 +1,169 @@
+{-# LANGUAGE ExistentialQuantification          #-}
+{-# LANGUAGE BangPatterns          #-}
+{-# LANGUAGE DeriveFunctor          #-}
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE DeriveGeneric #-}
+
+-- |
+-- Module      : Streamly.Parse.Types
+-- Copyright   : (c) 2019 Composewell Technologies
+-- License     : BSD3
+-- Maintainer  : streamly@composewell.com
+-- Stability   : experimental
+-- Portability : GHC
+--
+-- Streaming and backtracking parsers.
+--
+-- A parser is a generalization of a classical left fold. A fold never fails,
+-- it always produces a result, it has a default initial value. A parser may
+-- terminate or fail even without producing any result.
+
+-- Differences between a terminating fold and a parser:
+--
+-- \* A Parser is a superset of terminating fold with failure and backtracking
+--   added.
+-- \* A fold may be upgraded to a parser but vice versa is not true.
+-- \* A fold is not a partial function, as it may never fail.
+-- \* A fold's Step result is simpler as it does not backtrack.
+-- \* A fold's result type is simpler as it does not fail and it does not return
+--   unused input count.
+--
+-- Parse and Zip (zipWith) is possible only if both the parses are unskipping
+-- or are guaranteed to have outputs in sync. Or else we will have to keep the
+-- input buffered and may have to end the zip when one of them finishes. This
+-- is simple if the type guarantees no skip e.g. in case of classical left
+-- folds.
+--
+-- Terminating folds can have Sequential and Distribute and zip style
+-- applicatives but cannot have Alternative.
+--
+-- Parsers can have Sequential Applicative and Alternative but not Distribute
+-- and zip style applicative.
+--
+-- We can distribute and scan/parse a stream using both folds and parsers and
+-- merge the resulting streams using different merge strategies (e.g.
+-- interleaving or serial).
+
+module Streamly.Internal.Data.Parse.Types
+    (
+      Step (..)
+    , Parse (..)
+    )
+where
+
+-- | The return type of a fold step. A fold is driven by a fold driver one step
+-- at a time, at any time the driver may @extract@ and examine the result of
+-- the fold. The fold may backtrack at any point, therefore, the driver holds
+-- the input in a backtracking buffer. The buffer grows or shrinks based on the
+-- return values of the fold step execution.
+--
+-- When a fold step is executed it generates a new state of the fold along with
+-- a command to the fold driver. The command tells the fold driver whether to
+-- keep the input for future backtracking or drop it. The constructors of
+-- 'Step' represent the commands to the driver.
+--
+-- /Internal/
+--
+data Step s =
+      Hold s     -- ^ Hold the current input in the buffer, the input is
+                 -- accepted tentatively, but the fold may later fail with the
+                 -- input remaining unused.
+    | Keep Int s -- ^ Keep @n@ most recent inputs, including the input of this
```

Comment:
Can you replace this with `Discard s`?
Can you give me an example where we would use Keep i s where i > 0?

Reply:
Turns out `Keep i s` is indeed needed.
Example: A parser that succeeds if atleast the last `n` inputs `(n >= 0 && n < i)` satisfy a certain condition. Such a parser would need to keep history.

Reply:
> Example: A parser that succeeds if at least the last n inputs (n >= 0 && n < i) satisfy a certain condition. Such a parser would need to keep history even after success.

This can be done while updating the state accordingly though.